src/inc/til/unicode.h

```diff
@@ -0,0 +1,164 @@
+    {
+        return (wch & 0xFC00) == 0xDC00;
+    }
+
+    // Verifies the beginning of the given UTF16 string and returns the first UTF16 sequence
+    // or U+FFFD otherwise. It's not really useful and at the time of writing only a
+    // single caller uses this. It's best to delete this if you read this comment.
+    constexpr std::wstring_view utf16_next(std::wstring_view wstr) noexcept
+    {
+        auto it = wstr.begin();
+        const auto end = wstr.end();
+        auto ptr = &details::UNICODE_REPLACEMENT;
+        size_t len = 1;
+
+        if (it != end)
+        {
+            const auto wch = *it;
+            ptr = &*it;
+
+            if (is_surrogate(wch))
+            {
+                ++it;
+                const auto wch2 = it != end ? *it : wchar_t{};
+                if (is_leading_surrogate(wch) && is_trailing_surrogate(wch2))
+                {
+                    len = 2;
+                    ++it;
+                }
+                else
+                {
+                    ptr = &details::UNICODE_REPLACEMENT;
+                }
+            }
+        }
+
+        return { ptr, len };
+    }
+
+    // Splits a UTF16 string into codepoints, yielding `wstring_view`s of UTF16 text. Use it as:
+    //   for (const auto& str : til::utf16_iterator{ input }) { ... }
+    struct utf16_iterator
+    {
+        struct sentinel
+        {
+        };
+
+        struct iterator
+        {
+            using iterator_category = std::forward_iterator_tag;
+            using value_type = std::wstring_view;
+            using reference = value_type&;
+            using pointer = value_type*;
+            using difference_type = std::ptrdiff_t;
+
+            explicit constexpr iterator(utf16_iterator& p) noexcept :
+                _iter{ p }
+            {
+            }
+
+            const value_type& operator*() const noexcept
+            {
+                return _iter.value();
+            }
+
+            iterator& operator++() noexcept
+            {
+                _iter._advance = true;
+                return *this;
+            }
+
+            bool operator!=(const sentinel&) const noexcept
+            {
+                return _iter.valid();
+            }
+
+        private:
+            utf16_iterator& _iter;
+        };
+
+        explicit constexpr utf16_iterator(std::wstring_view wstr) noexcept :
+            _it{ wstr.begin() }, _end{ wstr.end() }, _advance{ _it != _end }
+        {
+        }
+
+        iterator begin() noexcept
+        {
+            return iterator{ *this };
+        }
+
+        sentinel end() noexcept
+        {
+            return sentinel{};
+        }
+
+    private:
+        bool valid() const noexcept
+        {
+            return _it != _end;
+        }
+
+        void advance() noexcept
+        {
+            const auto wch = *_it;
+            auto ptr = &*_it;
+            size_t len = 1;
+
+            ++_it;
+
+            if (is_surrogate(wch))
```

Comment:
The old code and more direct approach looks like this:
```cpp
if (is_leading_surrogate(wch)) { /* check if good */ }
else if (is_trailing_surrogate(wch)) { /* it's bad */ }
else { /* it's trivial */ }
```

Most text is in the BMP / has no surrogate pairs and so the last branch will be the most active one, but thanks to that the code is also easy to predict for the CPU and so it runs pretty fast. But we can do better by optimizing the code given our knowledge that surrogate pairs are seldom: The new code always first assumes that the current characters is not a surrogate pair, and then only then checks if it looks like one and if it's a valid one too. This leads to the redundant check of `is_surrogate` and then once again with `is_leading/trailing_surrogate`, but it'll run significantly faster that way.